src/mempool/order_state_watcher.ts

```diff
@@ -0,0 +1,150 @@
+import * as _ from 'lodash';
+import {schemas} from '0x-json-schemas';
+import {ZeroEx} from '../';
+import {EventWatcher} from './event_watcher';
+import {assert} from '../utils/assert';
+import {utils} from '../utils/utils';
+import {artifacts} from '../artifacts';
+import {AbiDecoder} from '../utils/abi_decoder';
+import {OrderStateUtils} from '../utils/order_state_utils';
+import {
+    LogEvent,
+    OrderState,
+    SignedOrder,
+    Web3Provider,
+    BlockParamLiteral,
+    LogWithDecodedArgs,
+    OnOrderStateChangeCallback,
+    ExchangeEvents,
+    TokenEvents,
+} from '../types';
+import {Web3Wrapper} from '../web3_wrapper';
+
+interface DependentOrderHashes {
+    [makerAddress: string]: {
+        [makerToken: string]: Set<string>,
+    };
+}
+
+interface OrderByOrderHash {
+    [orderHash: string]: SignedOrder;
+}
+
+export class OrderStateWatcher {
+    private _orders: OrderByOrderHash;
+    private _dependentOrderHashes: DependentOrderHashes;
+    private _web3Wrapper: Web3Wrapper;
+    private _callbackAsync?: OnOrderStateChangeCallback;
```

Comment:
is this async?

Reply:
It can be either but calling `await` on a sync function resolves immediately. It's much more problematic if we think the method is synchronous when it's actually async.